### PR TITLE
Add assign_(coord/mask/attr) methods of data_array

### LINKED ISSUE #2960: Add methods for adding coords/masks/attrs returning a modified object
Additional task from https://github.com/scipp/scipp/issues/2909

https://github.com/scipp/scipp/issues/2909#issue-1504298389
> Similar to https://docs.xarray.dev/en/stable/generated/xarray.DataArray.drop_vars.html.
> 
> To decide:
> 
>     Naming. drop_coords? drop_masks? etc.?

`drop_(coords/masks/attrs)` methods are implemented but we also want to have `add_(coords/masks/attrs)` methods.

### PR DESCRIPTION
**Pull request by YooSunYoung**
Fixes #2960 

TODO: python binding is not implemented yet

I used the name `assign` since I thought it should also be able to update the existing one.
Like this: https://docs.xarray.dev/en/stable/generated/xarray.DataArray.assign_coords.html

In the xarray, the syntax is like below
```
values = [...]
coord_name = "coord0"
dimension_name = "dim0"
assign_coords(coord_name=(dimension_name, values))
```

But since we are using the `Variable` which already has dimension name, it should be like below...?
```
coord_name = "coord0"
coord0 = sc.Variable(dims=["dim0"], values=[...])
da.assign_coords(coord_name=coord0)
```

### CODE REVIEW

**Comment by YooSunYoung** [comment]
I have a question!

Should we check if the given `coords/masks/attrs` has overlapping keys?

**Comment by SimonHeybrock** [comment]
> TODO: python binding is not implemented yet

Any reason why we don't implement the entire thing in Python?

**Comment by SimonHeybrock** [comment]
> I have a question!
> 
> Should we check if the given `coords/masks/attrs` has overlapping keys?

You mean if we are replacing existing coords? I think I would not check. After all, `__setitem__` does not check either?

**Comment by SimonHeybrock** [comment]
> > > TODO: python binding is not implemented yet
> > 
> > 
> > Any reason why we don't implement the entire thing in Python?
> 
> No... drop_* were in cpp so I just kept there together. We can write all of them in python, sure.

👍

**Comment by YooSunYoung** [comment]
> > I have a question!
> > Should we check if the given `coords/masks/attrs` has overlapping keys?
> 
> You mean if we are replacing existing coords? I think I would not check. After all, `__setitem__` does not check either?

No, I meant when you want to assign multiple of them, for example
```
assign_coords({'a': coord0, 'b':coord1, 'a': coord2})
```

**Comment by SimonHeybrock** [comment]
> > > I have a question!
> > > Should we check if the given `coords/masks/attrs` has overlapping keys?
> > 
> > 
> > You mean if we are replacing existing coords? I think I would not check. After all, `__setitem__` does not check either?
> 
> No, I meant when you want to assign multiple of them, for example
> 
> ```
> assign_coords({'a': coord0, 'b':coord1, 'a': coord2})
> ```

This will never reach `assign_coords` anyway, because Python first creates a `dict`?

**Comment by YooSunYoung** [comment]
> > > > I have a question!
> > > > Should we check if the given `coords/masks/attrs` has overlapping keys?
> > > 
> > > 
> > > You mean if we are replacing existing coords? I think I would not check. After all, `__setitem__` does not check either?
> > 
> > 
> > No, I meant when you want to assign multiple of them, for example
> > ```
> > assign_coords({'a': coord0, 'b':coord1, 'a': coord2})
> > ```
> 
> This will never reach `assign_coords` anyway, because Python first creates a `dict`?

In this case below it will survive but the keyword arguments will be used. I guess that's fine...? Or should we allow only one of them...?

```
assign_coords({'a': coord0}, a=coord0)
```

File: src/scipp/core/assignments.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    """
+
+    if coords is not None:
+        collected_coords = {**coords, **coords_kwargs}
+    else:
+        collected_coords = coords_kwargs
+
+    for coord_key, coord in collected_coords.items():
+        self.coords[coord_key] = coord
```
Comment (SimonHeybrock) [inline src/scipp/core/assignments.py:39]: We should not change the input. Must make a shallow copy first (`self = self.copy(deep=False)`)

File: tests/data_array_test.py [hunk 2]
```diff
@@ -476,3 +476,102 @@ def test_masks_update_from_dict_adds_items():
...
     assert sc.identical(da.masks['a'], sc.scalar(True))
     assert sc.identical(da.masks['b'], sc.scalar(False))
+
+
+def test_assign_coords():
+    data = sc.array(dims=['x', 'y', 'z'], values=np.random.rand(4, 3, 5))
+    da = sc.DataArray(data)
+    coord0 = sc.linspace('x', start=0.2, stop=1.61, num=4)
+    coord1 = sc.linspace('y', start=1, stop=4, num=3)
+    da.assign_coords({'coord0': coord0, 'coord1': coord1})
```
Comment (SimonHeybrock) [inline tests/data_array_test.py:486]: `da` should be kept unchanged. Make sure to also a a test for that.

File: src/scipp/core/assignments.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+from __future__ import annotations
+
+from typing import Dict, Union
+
+from ..typing import DataArray, Dataset
+
+
+def assign_coords(self, coords: Dict) -> Union[DataArray, Dataset]:
```
Comment (SimonHeybrock) [inline src/scipp/core/assignments.py:11]: Why did you remove the keyword-arg syntax?

Comment (YooSunYoung) [reply src/scipp/core/assignments.py:11]: > `coords` should be position-only, such that it does not prevent the use of a keyword arg of that name.

Sorry I misunderstood this comment. Then... what do you mean by `position-only` here ...?

Comment (SimonHeybrock) [reply src/scipp/core/assignments.py:11]: I meant the dict should be position-only, but keep the keyword args.

Comment (SimonHeybrock) [reply src/scipp/core/assignments.py:11]:
```python
def assign_coords(self, /, coords: Dict[str, Variable], **kwargs):
```
or something like that

Comment (YooSunYoung) [reply src/scipp/core/assignments.py]: Ah, I see. My question was more like, should we allow only either of them

More specifically, will this be allowed?
```
da.assign_coords({'coord0': coord0_a}, coord0=coord0_b)
```
Currently, the keyword argument will overwrite the positional argument in this case.

**Comment by YooSunYoung** [comment]
`rename_dims` allows using both positional arguments and keyword arguments,
but check if there is any overlapping names as you said,
so I did the similar thing for `assign_*`.

File: src/scipp/core/assignments.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    coords_posarg = {} if coords is None else coords
+    collected_coords = {**coords_posarg, **coords_kwargs}
+
+    if len(collected_coords) != len(coords_posarg) + len(coords_kwargs):
+        overlapped = set(coords).intersection(coords_kwargs)
+        raise ValueError(
+            'The names of coords passed in the dict '
+            'and as keyword arguments must be distinct.'
+            f'Following names were used in both places: {overlapped}.'
+        )
```
[the remark this answers is not in the document]
Comment (YooSunYoung) [reply src/scipp/core/assignments.py:39]: I can write a general function that combines positional dictionary argument and keyword arguments I guess...?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/core/assignments.py:39]: Looks like it exists already: https://github.com/scipp/scipp/blob/62778d23843b4f5cf1fd7e5e00ddfc201eea721a/src/scipp/core/dimensions.py#L12-L21

File: src/scipp/core/assignments.py [hunk 5]
```diff
@@ -8,6 +8,21 @@
 from ..typing import DataArray, Dataset, Variable
 
 
+def _combine_args(arg0: Optional[Dict[str, Variable]] = None, /, **kwargs):
```
Comment (SimonHeybrock) [inline src/scipp/core/assignments.py:11]: Still duplicate with https://github.com/scipp/scipp/blob/62778d23843b4f5cf1fd7e5e00ddfc201eea721a/src/scipp/core/dimensions.py#L12-L21. You can keep the error and not worry about it (since it should be extremely rare), or adjust it there to only print the keys?

File: src/scipp/core/argument_handlers.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+from typing import Dict, Optional
+
+
+def combine_dict_args(arg: Optional[Dict], /, **kwargs):
```
Comment (SimonHeybrock) [inline src/scipp/core/argument_handlers.py:7]: type hints missing now

**Review by SimonHeybrock — APPROVED** [review]
